util/src/journaldb/overlayrecentdb.rs

```diff
@@ -430,8 +434,51 @@ impl HashDB for OverlayRecentDB {
 		v.or_else(|| self.payload(key))
 	}
 
+	fn get_exec<'this>(
+		&'this self,
+		key: &'this H256,
+		f: &'this mut for<'a: 'this> FnMut(&'a [u8]),
+	) {
+		use ::std::cell::Cell;
+
+		let k = self.transaction_overlay.raw_ref(key);
+
+		if let Some((d, rc)) = k {
+			if rc > 0 {
+				f(d.as_ref());
+				return;
+			}
+		}
+
+		let was_called = Cell::new(false);
+
+		// This type annotation is necessary, the inferred lifetime is too restrictive otherwise.
+		// Possible bug in the Rust compiler.
+		let mut wrapper = |d: &[u8]| {
+			was_called.set(true);
+			f(d);
```

Comment:
Agreed, but importantly the `'this` lifetime is not the full lifetime of the borrow that calls it. It can be as short as you like as long as the inequalities are satisfied, and there's nothing that says that anything has to strictly outlive anything else.

Honestly, I trust the compiler on this one. When I tried to do a universal quantification or a parameterised concrete lifetime it failed, the only one it would accept was a bounded forall quantification. If you can write up an example of unsoundness caused by this then I'd be happy to submit it to the lang team as an unsoundness bug, but certainly as far as my (possibly faulty) intuition on lifetimes go this makes sense.